### PR TITLE
Implement slow-path PodTopologySpread logic without preFilterState

### LINKED ISSUE #87678: Calculate "preFilterState" for PodTopologySpread if it's not pre-calculated in preFilter plugin
Part of #87662.

In alpha phase, pod topology spread only works in defaultProvider, to promote it in beta, we should implement the slow-path preState calculation when it's not set:

https://github.com/kubernetes/kubernetes/blob/cad44602e1188a1feff0bc085dc8286c6c5d3640/pkg/scheduler/framework/plugins/podtopologyspread/filtering.go#L296

/sig scheduling
/kind feature
/assign

### PR DESCRIPTION
**Pull request by Huang-Wei**
**What type of PR is this?**

/kind feature
/sig scheduling

**What this PR does / why we need it**:

Implement the slow-path PodTopologySpread logic when preFilterState is absent. This is needed for CA (cluster autoscaler) so as to qualify this feature to be promoted to beta.

Unlike the preFilterState logic, when it comes to the slow-path, it means Filter() is called on every filteredNode and no calculation result is shared, so the logic is implemented differently, with a simplified data structure.

**Which issue(s) this PR fixes**:

Fixes #87678

**Special notes for your reviewer**:

I assume in CA's perspective, only the filter plugin needs to implement the slow-path logic. Let me know if other case (preemption, score) is also needed.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/cc @ahg-g @alculquicondor

### CODE REVIEW

**Comment by haosdent** [comment]
+1

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 1]
```diff
@@ -280,6 +280,79 @@ func calPreFilterState(pod *v1.Pod, allNodes []*schedulernodeinfo.NodeInfo) (*pr
...
+		// The node won't be a fit if it doesn't have c.topologyKey present.
+		tpKey := c.topologyKey
+		tpVal, ok := nodeInfo.Node().Labels[tpKey]
+		if !ok {
+			klog.V(5).Infof("node '%s' doesn't have required label '%s'", nodeInfo.Node().Name, tpKey)
+			return framework.NewStatus(framework.Unschedulable, ErrReasonConstraintsNotMatch)
+		}
+
+		tpPairToMatchNum := make(map[topologyPair]int32)
+		// Iterate through all nodes.
```
Comment (haosdent) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go]: Actually should we move 

```
nodeInfos, _ := pl.sharedLister.NodeInfos().List()
		for _, nodeInfo := range nodeInfos {
			// (1) Ensure the node matches the incoming Pod's node selector/affinity, if specified.
			if !pluginhelper.PodMatchesNodeSelectorAndAffinityTerms(pod, nodeInfo.Node()) {
				continue
			}
```

and

```
for _, existingPod := range nodeInfo.Pods() {
				if existingPod.Namespace != pod.Namespace {
					continue
				}
```

out of `for _, c := range constraints` to get a pre-filtered pod list first. Suppose has huge node and several constraints, time complexity would increase a lot due to call `pluginhelper.PodMatchesNodeSelectorAndAffinityTerms(pod, nodeInfo.Node()` every time.

Comment (Huang-Wei) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering.go]: Calculating pre-filtered pod list doesn't look right as you can't only focus on matching Pods, instead, you need to also consider the case that nodes can be legit and don't have any matching Pod. In that case, the pre-filtered pod list lost that info.

I **_was_** thinking about calculating a pre-filtered node list, but it's only beneficial when len(constraints) >= 2, and given that it's a slow-path which we don't expect it to be executed by critical paths, so I gave up this idea.

Comment (haosdent) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering.go]: got it, thx for detail explaination.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 2]
```diff
@@ -280,6 +280,78 @@ func calPreFilterState(pod *v1.Pod, allNodes []*schedulernodeinfo.NodeInfo) (*pr
 	return &s, nil
 }
 
+// fit checks if pod is a good fit on the given node.
+func (pl *PodTopologySpread) fit(pod *v1.Pod, nodeInfo *nodeinfo.NodeInfo) *framework.Status {
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go]: What is the benefit of this versus calling `calPreFilterState`? At a quick glance, complexity looks the same.

**Review by ahg-g — COMMENTED** [review]
Thanks Wei, looks good to me.

I wonder if what CA actually needs is just a method to add/remove a node from prefilter state, and that can potentially be a lot faster for CA to simulate adding/removing nodes than computing the whole state all over again for each node.

/cc @MaciekPytel @losipiuk 
adding CA folks for input

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 3]
```diff
@@ -280,6 +280,78 @@ func calPreFilterState(pod *v1.Pod, allNodes []*schedulernodeinfo.NodeInfo) (*pr
...
+		// 'existing matching num' + 'if self-match (1 or 0)' - 'global min matching num' <= 'maxSkew'
+		pair := topologyPair{key: tpKey, value: tpVal}
+		matchNum := tpPairToMatchNum[pair]
+		selfMatchNum := int32(0)
+		if c.selector.Matches(podLabelSet) {
+			selfMatchNum = 1
+		}
+		skew := matchNum + selfMatchNum - minMatchNum
+		if skew > c.maxSkew {
+			return framework.NewStatus(framework.Unschedulable, ErrReasonConstraintsNotMatch)
```
Comment (notpad) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go]: Should we add a `klog.V(5).Infof` here to be consistent with other returns?

**Comment by losipiuk** [comment]
> Thanks Wei, looks good to me.
> 
> I wonder if what CA actually needs is just a method to add/remove a node from prefilter state, and that can potentially be a lot faster for CA to simulate adding/removing nodes than computing the whole state all over again for each node.
> 
> /cc @MaciekPytel Łukasz Osipiuk
> adding CA folks for input

Thanks for work and looping us in!

With current implementation of CA based on scheduler framework (currently in review: https://github.com/kubernetes/autoscaler/pull/2709), we are always calling all pre-filter plugins before filters. So I do not believe this code path will trigger at all. We are currently not doing any optimizations to limit amount of work spent in scheduler code. For each Pod, which we are simulating scheduling for, we are calling out both pre-filters and filters ([link](https://github.com/losipiuk/autoscaler/blob/lo/scheduler-framework-poc2/cluster-autoscaler/simulator/scheduler_based_predicates_checker.go#L224-L265)). Only optimization we have is that if we want to check multiple nodes in a loop we are calling to pre-filters once and then filters for each node ([link](https://github.com/losipiuk/autoscaler/blob/lo/scheduler-framework-poc2/cluster-autoscaler/simulator/scheduler_based_predicates_checker.go#L168-L214)).

We are not using methods which mutate `CycleState` at all right now. We might add those do not fit well in our code flow, and scale test so far showed that performance is good enough. Not worse that 1.16.

Do you guys expect that performance will get substantially worse with our approach, with equal-pods spreading enabled?

**Comment by ahg-g** [comment]
> > Thanks Wei, looks good to me.
> > I wonder if what CA actually needs is just a method to add/remove a node from prefilter state, and that can potentially be a lot faster for CA to simulate adding/removing nodes than computing the whole state all over again for each node.
> > /cc @MaciekPytel Łukasz Osipiuk
> > adding CA folks for input
> 
> Thanks for work and looping us in!
> 
> With current implementation of CA based on scheduler framework (currently in review: [kubernetes/autoscaler#2709](https://github.com/kubernetes/autoscaler/pull/2709)), we are always calling all pre-filter plugins before filters. So I do not believe this code path will trigger at all. We are currently not doing any optimizations to limit amount of work spent in scheduler code. For each Pod, which we are simulating scheduling for, we are calling out both pre-filters and filters ([link](https://github.com/losipiuk/autoscaler/blob/lo/scheduler-framework-poc2/cluster-autoscaler/simulator/scheduler_based_predicates_checker.go#L224-L265)). Only optimization we have, is that if we want to check multiple nodes in a loop, we are calling to pre-filters once and then filters for each node ([link](https://github.com/losipiuk/autoscaler/blob/lo/scheduler-framework-poc2/cluster-autoscaler/simulator/scheduler_based_predicates_checker.go#L168-L214)).
> 
> We are not using methods which mutate `CycleState` at all right now. We might add those in the future. Yet they do not fit well in our code flow, and scale test so far showed that performance is good enough. Not worse that 1.16.
> 
> Do you guys expect that performance will get substantially worse with our approach, with equal-pods spreading enabled?

Thanks Lukasz!

I am not sure what your old approach, so I can't comment if it will get worse, but calling prefilters once and then filter on all nodes makes a lot of sense, that is what we do in the scheduler.

If your new approach is to call PreFilters always, then we would like to cleanup the similar logic in multiple other plugins (which is there only for CA). I will open an issue for that and cc you.

**Comment by ahg-g** [comment]
/hold

since we may not need this after all.

**Comment by losipiuk** [comment]
> > > Thanks Wei, looks good to me.
> > > I wonder if what CA actually needs is just a method to add/remove a node from prefilter state, and that can potentially be a lot faster for CA to simulate adding/removing nodes than computing the whole state all over again for each node.
> > > /cc @MaciekPytel Łukasz Osipiuk
> > > adding CA folks for input
> > 
> > 
> > Thanks for work and looping us in!
> > With current implementation of CA based on scheduler framework (currently in review: [kubernetes/autoscaler#2709](https://github.com/kubernetes/autoscaler/pull/2709)), we are always calling all pre-filter plugins before filters. So I do not believe this code path will trigger at all. We are currently not doing any optimizations to limit amount of work spent in scheduler code. For each Pod, which we are simulating scheduling for, we are calling out both pre-filters and filters ([link](https://github.com/losipiuk/autoscaler/blob/lo/scheduler-framework-poc2/cluster-autoscaler/simulator/scheduler_based_predicates_checker.go#L224-L265)). Only optimization we have, is that if we want to check multiple nodes in a loop, we are calling to pre-filters once and then filters for each node ([link](https://github.com/losipiuk/autoscaler/blob/lo/scheduler-framework-poc2/cluster-autoscaler/simulator/scheduler_based_predicates_checker.go#L168-L214)).
> > We are not using methods which mutate `CycleState` at all right now. We might add those in the future. Yet they do not fit well in our code flow, and scale test so far showed that performance is good enough. Not worse that 1.16.
> > Do you guys expect that performance will get substantially worse with our approach, with equal-pods spreading enabled?
> 
> Thanks Lukasz!
> 
> I am not sure what your old approach, so I can't comment if it will get worse, but calling prefilters once and then filter on all nodes makes a lot of sense, that is what we do in the scheduler.

By old approach I meant the CA versions which were not based on scheduler framework. Then we called out to predicate functions directly and only computed predicate meta (which got replaced by pre-filters now) in subset of flows (almost never).

For scheduler framework based CA we always call out to pre-filters (as we kinda agreed on that before implementation :) ). 

And my question about regression was if performance will get worse when `PodTopologySpread` gets promoted to beta. I assumed that pre-filter related to that functionality is not present now, in the set of plugins registered by default. Am I right?

> If your new approach is to call PreFilters always, then we would like to cleanup the similar logic in multiple other plugins (which is there only for CA). I will open an issue for that and cc you.

Thx.

**Comment by Huang-Wei** [comment]
Thanks @losipiuk  for explaining the CA usecases in detail.

This PR is only useful if CA wants to call the Filter plugin individually. If you call multiple nodes in batch, then involving the preFilter is a better fit. 

> And my question about regression was if performance will get worse when PodTopologySpread gets promoted to beta. I assumed that pre-filter related to that functionality is not present now, in the set of plugins registered by default. Am I right?

Nope :) We've got the preFilter implemented since alpha (1.16), as it's mandatory for plugins such as PodTopologySpread which needs to calculate the pods across nodes.
However, one thing _maybe_ worth noting is that, in the PreFilter function, the data structure of preState (we called it predicate metadata before) is designed for both Filter and Preemption - the part for Preemption may not be necessary for CA I suppose. In other words, if we don't consider Preemption, the memory print of the preState can technically be optimized a bit, for CA only, but the time complexity should be the same level.

**Comment by ahg-g** [comment]
I think in the past metadata calculation overhead wasn't zero for unused features, this has changed.

**Comment by Huang-Wei** [comment]
Closing this as we don't need keep the slow-path logic now.

**Pull request closed without merging**